Complete Upgrade: Babel 6 / React 0.14.x / Node 5.4/ react-transform-hmr / all dependencies are up to date

Bartekus opened the pull request:
Surface Upgrade Completed:
 - Babel 6 
 - React 0.14.x 
 - Node 5.4 
 - react-hot-loader (deprecated) replaced with react-transform-hmr

##Webpack Modification (WIP)

 - webpack.config.coverage.js 
> "isparta?{ babel: { stage: 2 } }" 
replaced with:
> "isparta?{babel}" 
not optimal, given that currently there is only one .babelrc config utilizing es2015, react & stage-1

 - webpack.config.hot.js
> ["react-hot", "babel-loader?optional=runtime&stage=2"]
replaced with:
> ["babel"]
not optimal as described above + address babel-runtime 

 - webpack.config.js
> ["babel-loader?optional=runtime&stage=2"]
replaced with:
> ["babel"]
critical since base webpack config is used in tests 

## Environment specific

> Node 0.10 / 0.12
replace with:
> Node 5.4

> NPM 2.1.x
replaced with:
> NPM 3.3.x

- Intended to bring Node / NPM to most current versions

TODO: 
Structural Code Modification needed on: 
/client/components/error-panel.jsx
/client/components/output-panel.jsx
/client/components/types.jsx

Currently not React 0.14.x compliant which results in: 
TypeError: ... Property value of JSXAttribute expected node to be of a type ['JSXElement','StringLiteral','JSXExpressionContainer'] but instead got 'CallExpression'

Bartekus commented:
Work-In-Progress - Submittal for reference as objections/suggestions are addressed.

ryan-roemer commented on webpack.config.coverage.js:
```diff
@@ -13,8 +13,8 @@ module.exports = _.merge({}, testCfg, {
-    ]
-  }
+        loader: "isparta?{babel}"
+      },
```
Please remove trailing commas.

ryan-roemer commented on test/client/spec/components/types-title.spec.jsx:
```diff
@@ -1,12 +1,12 @@
  */
-import React from "react/addons";
+import React from "react";
+import TestUtils from 'react-addons-test-utils';
```
GLOBAL: Please use double quotes.

ryan-roemer commented on appveyor.yml:
```diff
@@ -5,10 +5,9 @@ environment:
   matrix:
-    - nodejs_version: 0.10
-    - nodejs_version: 0.12
+    - nodejs_version: 5.4
```
Let's leave at least `0.12` for continued testing and maybe add `4.stable` too

ryan-roemer commented on package.json:
```diff
@@ -65,75 +65,87 @@
+    "nodemon": "^1.8.1",
+    "phantomjs": "^1.9.19",
+    "react-addons-test-utils": "^0.14.5",
+    "react-jasmine-matchers": "^2.0.0",
```
What are `jasmine-matchers` for?

ryan-roemer commented on webpack.config.hot.js:
```diff
@@ -18,7 +18,7 @@ module.exports = {
         test: /\.jsx?$/,
         exclude: [/node_modules/],
-        loaders: ["react-hot", "babel-loader?optional=runtime&stage=2"]
+        loaders: ["babel"]
```
How does `hot` loading work now?

Ideally `webpack.config.hot.js` gives us a build _with_ hot reloading and `webpack.config.dev.js` (based off `webpack.config.js`) gives us a build _without_ hot reloading.

ryan-roemer commented:
@Bartekus -- Thanks for the initial WIP work! I've got a few preliminary comments.

Also, I think the `eslint-*` stuff may need updates for modern configs / rules. Also, the `fs.access` is a library depending now on `0.12+` node, so you may need to pin that library specifically to an earlier version.

Ultimately, get CI cleaned up such that it correctly runs on your machine. The last stuff with Karma through Sauce _will_ fail because Travis won't allow you to use this repo's secret credentials via a forked PR (I'll hook up that last part when we're at that stage of the process).

Bartekus commented:
- I've recombined all webpack configs into one and adjusted package.json to utilize 
```
NODE_ENV=/*mode*/
```
in order to trigger them.

I've also removed .babelrc and moved the setting to webpack.config.

 - Currently karma has been separated from the webpack however it's my intention to have only one config file hence I'll be merging it into back into webpack.config.js and then refactor it to make it concise and clean. As with webpack, the NODE_ENV will be used to trigger appropriate build.
 
 - I've adjusted and corrected all errors but TypeError in:
/client/components/error-panel.jsx
/client/components/output-panel.jsx
/client/components/types.jsx
```
Property value of JSXAttribute expected node to be of a type ['JSXElement','StringLiteral','JSXExpressionContainer'] but instead got 'CallExpression': 
```
These are the last 3 errors (+ 2 warnings) present that I can spot by npm run check && npm run dev
I'd have tried to fix them by adding constructor/super but so far I have not had any success thus I'll require some pointers to bring this into finalization.

As to linting standards:
https://github.com/feross/standard
or
https://github.com/Flet/semistandard
would provide solid base.
Personally I'm for the later, since I do like ;

ryan-roemer commented:
@Bartekus -- I think it's perhaps better to focus on **just** the upgrade. Offhand, a couple of things to keep in mind:

* We're going to use the existing eslint configs, but updated to modern https://github.com/walmartlabs/eslint-config-defaults/tree/master/configurations/walmart -- we help maintain these and have deep experience and familiarity with these over many, many projects.
* Please keep the webpack configs separate from the karma configs. This enables much better workflows that have webpack dev processes running in a separate process from the karma one. And in large file projects, the webpack build time is non-trivial.
* Very soon, we are going to be moving all the eslint, karma, webpack, and other configs to a [builder](https://github.com/FormidableLabs/builder) archetype (analogous to [`builder-react-component](https://github.com/FormidableLabs/builder-react-component) archetype, but for a full react-based application server) for better management of dependencies, workflows and honing down all the boilerplate in this specific project.

Ideally, the best approach, infrastructure-wise is "change the structure of files" + "touch things" as little as possible.

ryan-roemer commented on appveyor.yml:
```diff
@@ -5,10 +5,11 @@ environment:
   matrix:
-    - nodejs_version: 0.10
     - nodejs_version: 0.12
+    - nodejs_version: 4.stable
```
Is this an actual version? (Sorry if I led you astray, I was giving pseudocode in earlier comment)

ryan-roemer commented on client/components/input.jsx:
```diff
@@ -6,13 +6,15 @@ import { connect } from "react-redux";
 import Input from "react-bootstrap/lib/Input";
 import { setConversionValue, fetchConversions } from "../actions/";
 
+const magicNum = 13;
```
`ENTER_KEY`

ryan-roemer commented on client/utils/api.js:
```diff
@@ -4,15 +4,17 @@
-  BASE_URL: "",
+const badRequest = 400;
+const BASE_URL = "http://";
+const COLON_URL = ":";
```
Do these really need to be constents above?

ryan-roemer commented on client/utils/api.js:
```diff
@@ -4,15 +4,17 @@
 import "isomorphic-fetch";
 
-const api = {
-  BASE_URL: "",
```
I'd like to keep `BASE_URL` exposed on the object.

ryan-roemer commented on package.json:
```diff
@@ -3,50 +3,50 @@
-    "test-client-cov": "node node_modules/karma/bin/karma start karma.conf.coverage.js",
-    "test-client-dev": "node node_modules/karma/bin/karma start karma.conf.dev.js",
-    "test-server-rest": "mocha --opts test/server/mocha.opts test/server/rest",
+    "lint-client":          "eslint --ext .js,.jsx -c .eslintrc-client client templates",
```
Please revert extra spacing.

ryan-roemer commented on test/client/spec/components/types-title.spec.jsx:
```diff
@@ -1,16 +1,18 @@
 // Use `TestUtils` to inject into DOM, simulate events, etc.
 // See: https://facebook.github.io/react/docs/test-utils.html
-const TestUtils = React.addons.TestUtils;
+const expectedNum = 3;
```
GLOBAL: Please do not move assertion / test local variables out of scope.

ryan-roemer commented on test/server/rest/api.spec.js:
```diff
@@ -4,6 +4,7 @@
 // See: http://visionmedia.github.io/superagent/
 // See: https://github.com/visionmedia/supertest
 var request = require("supertest");
+var httpOk = 200;
```
GLOBAL: Please move back to test local scope.

ryan-roemer commented on test/server/rest/base.spec.js:
```diff
@@ -13,7 +13,8 @@ process.env.NODE_ENV = process.env.NODE_ENV || "test-rest-spec";
 var sinon = require("sinon");
 var app = require("../../../server");
-var PORT = process.env.TEST_REST_PORT || 3020;
+var testPort = 3020;
```
GLOBAL: If you have a top level variable that isn't going to change after initialization, use `CAPS_CASE`

ryan-roemer replied:
Also, switch this one back to the way it was -- there's no readability benenfit to having another variables

ryan-roemer commented on karma.conf.js:
```diff
@@ -6,7 +6,79 @@
  * the test files one-off for just a single run. This is appropriate for a
  * CI environment or if you're not otherwise running `npm run dev|hot`.
  */
-var webpackCfg = require("./webpack.config.test");
```
GLOBAL: Please try and undo all this copy and paste stuff and refactor to reuse webpack configs and share across karma like it was before. Thanks!

ryan-roemer commented on package.json:
```diff
@@ -3,50 +3,50 @@
+    "server":               "nodemon --watch client --watch server --watch templates --ext js,jsx server/index.js",
+    "server-dev":           "nodemon --watch client --watch server --watch templates --ext js,jsx server/index-dev.js",
+    "server-hot":           "nodemon --watch client --watch server --watch templates --ext js,jsx server/index-hot.js",
+    "server-wds-dev":       "NODE_ENV=development && webpack-dev-server --progress --colors --port 2992",
```
All code in this repo needs to be windows / mac / linux compatible. This code won't work on windows.

ryan-roemer commented on package.json:
```diff
@@ -3,50 +3,50 @@
   "version": "0.0.1",
   "description": "Converter application (React)",
   "scripts": {
```
GLOBAL: Please try to _not change_ any of the workflow tasks / limit changes to the minimum absolutely necessary. Another developer is already in the process of extracting the existing workflows to `builder`, and we want to limit churn there + the existing set is mostly what we'll want to keep targeting.
